### ISSUE TITLE
Thoughts on Interface Design

### ISSUE THREAD

**Issue by Non-Contradiction** [Non-Contradiction/autodiffr#4]
## Dimensions
The package need to deal with
* derivatives and second-order (or higher-order) derivatives of R1 -> R1 functions
* gradients and hessians of Rn(n>1) -> R1 functions
* jacobians of Rn(n>1) -> Rm(m>1) functions

And the package will provide separate functions, like `deriv` for derivatives,
and `grad` for gradients.

It is expected that users understand the difference and use corresponding functions when dealing with problems in different dimensionality.

Maybe we can include some error messages like "`grad` is not for R1 -> R1 functions, maybe you want to use `deriv`?"
if the user uses `grad` for a scalar function of one variable.

## Arguments
The arguments to the interface functions will look like:
`grad(func, x = NULL, mode = c("forward", "reverse"), ...)
`
where
* `func` is the original function.
* `x` is where the gradient is calculated, and it should be the first argument to `func` except the arguments matched by `...`. If it is `NULL`, then a gradient function will be returned.
* `mode` is whether forward or reverse mode automatic differentiation is used,
* And `...` are other arguments to the function `func`.

**Comment by nashjc** [comment]
If possible, it would be better if grad() could recognize x is scalar. The grad() function of numDeriv seems to work OK

Example:
tfn <- function(x) {x^2 - x^3}
efn <- expression( x^2 - x^3 )
dfn <- D(efn, "x")
dfn
library(numDeriv)
x <- 2
grad(tfn,x)
eval(dfn)
x<- 1)
grad(tfn,x)
eval(dfn)
x<- -1
grad(tfn,x)
eval(dfn)
x<- -3
grad(tfn,x)
eval(dfn)

cat("\nNOW VECTORS\n")
x <- c(2,2)
grad(tfn,x)
eval(dfn)
x<- c(1,2,3)
grad(tfn,x)
eval(dfn)

**Comment by Non-Contradiction** [comment]
It seems to be possible, just call `forwrad.derivative` when `x` is a scalar and `forward.grad` or `reverse.grad` when `x` is a vector.

Should we do similar things with higher-order like a hypothetical `hessian`? So if `x` is a scalar and `f(x)` is a scalar too, then just return the second order derivative? If `x` is a scalar and `f(x)` is a vector, return the jacobian of the first derivative??

**Comment by nashjc** [comment]
If possible, we should make our codes "just work" because most R users don't even know what a gradient is. Seriously, I have had quite advanced researchers who were not in the math/stat field ask that. Moreover, we may call the functions from within optim, optimr or other functions that users don't actually see.

**Comment by Non-Contradiction** [comment]
Yes, I think it's totally reasonable to do this, and we can even make `deriv` and `grad` the same function. And it's also reasonable to make `hessian` work for scalar `x` and scalar `f(x)`.

But the thing that if `x` is a scalar and `f(x)` is a vector, returns the jacobian of the first derivative is a little weird from my point of view and it may not be a good idea to do this because it can cause silent mistakes.

So maybe just the way in the first paragraph of this comment, make `deriv` and `grad` the same function, and make `hessian` work for scalar `x` and scalar `f(x)`?

**Comment by Non-Contradiction** [comment]
BTW, I updated the first post with some new thought on arguments of the high-level interface APIs.

The default options are for the most general case without any assumption. But in cases like `optim` and `optimr`, maybe it is more reasonable to have another default, particular with `xsize` set.

Is there any mechanism for `optim` and `optimr` to handle such things? Or should we have some specific version for these packages?

**Comment by Non-Contradiction** [comment]
Currently, there are functions `deriv`, `grad`, `jacobian` and `hessian`. The function interface is similar to the one in the update of first post.

All the functions `deriv`, `grad`, `jacobian` and `hessian` can deal with functions with either scalar or vector input with length greater than 1.

`deriv`, `grad` are for first order derivatives of functions with scalar output (in fact, `deriv` and `grad` are the same) 
and `jacobian` are for first order derivatives of functions with vector output with length greater or equal than 1.
In fact, it can be seen that `jacobian` can deal with all the functions that `deriv` and `grad` can deal with, but pay attention to that `jacobian` will always return the result as a Jacobian matrix.

`hessian` is for second order derivatives of functions with scalar output.
